crates/ruff_db/src/file_system.rs

```diff
@@ -0,0 +1,265 @@
+use std::fmt::Formatter;
+use std::ops::Deref;
+use std::path::Path;
+
+use camino::{Utf8Path, Utf8PathBuf};
+use filetime::FileTime;
+
+pub use memory::MemoryFileSystem;
+pub use os::OsFileSystem;
+
+mod memory;
+mod os;
+
+pub type Result<T> = std::io::Result<T>;
+
+/// A file system that can be used to read and write files.
+///
+/// The file system is agnostic to the actual storage medium, it could be a real file system, a combination
+/// of a real file system and an in-memory file system in the case of an LSP where unsaved changes are stored in memory,
+/// or an all in-memory file system for testing.
+pub trait FileSystem {
+    /// Reads the metadata of the file or directory at `path`.
+    fn metadata(&self, path: &FileSystemPath) -> Result<Metadata>;
+
+    /// Reads the content of the file at `path`.
+    fn read(&self, path: &FileSystemPath) -> Result<String>;
+
+    /// Returns `true` if `path` exists.
+    fn exists(&self, path: &FileSystemPath) -> bool;
+}
+
+// TODO support untitled files for the LSP use case. Wrap a `str` and `String`
+//    The main question is how `as_std_path` would work for untitled files, that can only exist in the LSP case
+//    but there's no compile time guarantee that a [`OsFileSystem`] never gets an untitled file path.
```

Comment:
CC: @snowsignal I don't plan to support this as part of this PR but something we need to figure out for red-knot/LSP

Reply:
Thanks for pinging me on this!